### PR TITLE
Adds Marlin mode

### PR DESCRIPTION
**Pull request by kobigurk**
The protocol is implemented as described here:  https://hackmd.io/glaJt8KuQE26WCoZ4wfFTw. It generates an SRS for SonicKZG10.

### CODE REVIEW

File: aleo-setup1/src/lib.rs [hunk 1]
```diff
@@ -26,3 +28,12 @@ pub fn curve_from_str(src: &str) -> Result<CurveKind, String> {
     };
     Ok(curve)
 }
+
+pub fn proving_system_from_str(src: &str) -> Result<ProvingSystem, String> {
+    let system = match src.to_lowercase().as_str() {
+        "groth16" => ProvingSystem::Groth16,
+        "marlin" => ProvingSystem::Marlin,
+        _ => return Err("unsupported proving system. Currently supported: groth16".to_string()),
```
[the remark this answers is not in the document]
Comment (kobigurk) [reply aleo-setup1/src/lib.rs:36]: Yep!

File: Cargo.toml [hunk 2]
```diff
@@ -27,3 +27,17 @@ opt-level = 0
 opt-level = 0
 debug-assertions = true
 debug = true
+
+
+#[patch.'ssh://git@github.com/AleoHQ/snarkOS.git']
```
[the remark this answers is not in the document]
Comment (kobigurk) [reply Cargo.toml:32]: Yep, done

File: phase1/src/helpers/accumulator.rs [hunk 3]
```diff
@@ -126,7 +129,10 @@ pub fn deserialize<E: PairingEngine>(
     let tau_g2 = in_tau_g2.read_batch(compressed, check_input_for_correctness)?;
     let alpha_g1 = in_alpha_g1.read_batch(compressed, check_input_for_correctness)?;
     let beta_g1 = in_beta_g1.read_batch(compressed, check_input_for_correctness)?;
-    let beta_g2 = (&*in_beta_g2).read_element(compressed, check_input_for_correctness)?;
+    let beta_g2 = match parameters.proving_system {
+        ProvingSystem::Groth16 => (&*in_beta_g2).read_element(compressed, check_input_for_correctness)?,
+        ProvingSystem::Marlin => E::G2Affine::prime_subgroup_generator(),
```
[the remark this answers is not in the document]
Comment (kobigurk) [reply phase1/src/helpers/accumulator.rs:134]: Yep, exactly.

File: phase1/Cargo.toml [hunk 4]
```diff
@@ -26,13 +26,21 @@ rand = { version = "0.7" }
...
 
 snarkos-curves = { git = "ssh://git@github.com/AleoHQ/snarkOS.git", package = "snarkos-curves", default-features = false }
 snarkos-models = { git = "ssh://git@github.com/AleoHQ/snarkOS.git", package = "snarkos-models", default-features = false }
 snarkos-utilities = { git = "ssh://git@github.com/AleoHQ/snarkOS.git", package = "snarkos-utilities" }
+snarkos-objects = { git = "ssh://git@github.com/AleoHQ/snarkOS.git", package = "snarkos-objects" }
+snarkos-errors = { git = "ssh://git@github.com/AleoHQ/snarkOS.git", package = "snarkos-errors" }
+snarkos-posw = { git = "ssh://git@github.com/AleoHQ/snarkOS.git", package = "snarkos-posw" }
+snarkos-algorithms = { git = "ssh://git@github.com/AleoHQ/snarkOS.git", package = "snarkos-algorithms" }
+snarkos-gadgets = { git = "ssh://git@github.com/AleoHQ/snarkOS.git", package = "snarkos-gadgets" }
+snarkos-polycommit = { git = "ssh://git@github.com/AleoHQ/snarkOS.git", package = "snarkos-polycommit" }
```
[the remark this answers is not in the document]
Comment (kobigurk) [reply phase1/Cargo.toml:43]: I'll try that

[the remark this answers is not in the document]
Comment (kobigurk) [reply phase1/Cargo.toml]: Seems like indeed posw at least is not setup correctly - it uses the errors crate without default-features = false

[the remark this answers is not in the document]
Comment (kobigurk) [reply phase1/Cargo.toml]: Let's leave that to a future PR?

**Review by howardwu — APPROVED** [review]

**Pull request merged**